res/controllers/Numark-Party-Mix.scripts.js

```diff
@@ -0,0 +1,440 @@
+        });
+    };
+
+    // stop the blink timer
+    this.blinkLedOff = function() {
+        if (this.blinkTimer === 0) {
+            return;
+        }
+
+        engine.stopTimer(this.blinkTimer);
+        this.blinkTimer = 0;
+    };
+
+    this.disablePadLights = function() {
+        for (var i = 0; i < 4; i++) {
+            midi.sendShortMsg(0x93 + deckNumber, 0x14 + i, ledOff);
+        }
+    };
+
+    this.currentMode = this.modes[NumarkPartyMix.PadModeControls.HOTCUE];
+};
+NumarkPartyMix.PadSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+
+NumarkPartyMix.ModeHotcue = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkPartyMix.PadModeControls.HOTCUE;
+
+    this.connections = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.connections[i] = new components.HotcueButton({
+            group: "[Channel" + deckNumber + "]",
+            midi: [0x93 + deckNumber, 0x14 + i],
+            number: i + 1,
+            outConnect: false
+        });
+    }
+};
+NumarkPartyMix.ModeHotcue.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkPartyMix.ModeLoop = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkPartyMix.PadModeControls.AUTOLOOP;
+
+    this.connections = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.connections[i] = new components.Button({
+            group: "[Channel" + deckNumber + "]",
+            midi: [0x93 + deckNumber, 0x14 + i],
+            size: NumarkPartyMix.autoLoopSizes[i],
+            inKey: "beatloop_" + NumarkPartyMix.autoLoopSizes[i] + "_toggle",
+            outKey: "beatloop_" + NumarkPartyMix.autoLoopSizes[i] + "_enabled",
+            outConnect: false
+        });
+    }
+};
+NumarkPartyMix.ModeLoop.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkPartyMix.ModeSampler = function(deckNumber) {
+    components.ComponentContainer.call(this);
+    this.control = NumarkPartyMix.PadModeControls.SAMPLER;
+    var sampleoffset = (deckNumber - 1)*4;
+    this.connections = new components.ComponentContainer();
+    for (var i = 0; i < 4; i++) {
+        this.connections[i] = new components.SamplerButton({
+            midi: [0x93 + deckNumber, 0x14 + i],
+            number: 1 + i + sampleoffset,
+            outConnect: false,
+            unshift: null,
+            input: function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
+                    if (engine.getValue(this.group, "track_loaded") === 0) {
+                        engine.setValue(this.group, "LoadSelectedTrack", 1);
+                    } else {
+                        engine.setValue(this.group, "start_play", 1);
+                    }
+                } else {
+                    engine.setValue(this.group, "start_stop", 1);
```

Comment:
so one has to hold the sampler button to keep it playing?
why not make it a toggle?
if (group.isPlaying) { stop } else { play }

I guess the behaviour depends on the use case / sample length. For typical one-shot samples (airhorn, high-hat, vocals) one press to trigger it suits probably better. The trigger finger can immediately be used for other stuff.
For longer samples that should not finish on their own press-and-hold may work better.
IIRC Mixxx mappings use both, trigger and hold.

Both cases could be handled with a longpress timer.
What do you think?

Reply:
yes makes sense (i don't use samplers)